crates/starknet-types-core/src/hash/poseidon.rs

```diff
@@ -0,0 +1,63 @@
+use crate::felt::Felt;
+use lambdaworks_crypto::hash::poseidon::{
+    starknet::PoseidonCairoStark252, Poseidon as PoseidonLambdaworks,
+};
+use lambdaworks_math::field::{
+    element::FieldElement, fields::fft_friendly::stark_252_prime_field::Stark252PrimeField,
+};
+
+use super::traits::StarkHash;
+
+pub struct Poseidon;
+
+impl StarkHash for Poseidon {
+    /// Computes the Poseidon hash of two Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Hashing/hash-functions/#poseidon_hash.>
+    fn hash(felt_0: &Felt, felt_1: &Felt) -> Felt {
+        Felt(PoseidonCairoStark252::hash(&felt_0.0, &felt_1.0))
+    }
+    /// Computes the Poseidon hash of an array of Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Cryptography/hash-functions/#poseidon_array_hash.>
+    fn hash_array(felts: &[Felt]) -> Felt {
+        // Non-copy but less dangerous than transmute
+        // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
+        Felt(PoseidonCairoStark252::hash_many(unsafe {
+            core::slice::from_raw_parts(
+                felts.as_ptr() as *const FieldElement<Stark252PrimeField>,
```

Comment:
I think for it to be entierly safe you need `#[repr(transparent)]` ontop of `Felt`

Reply:
it is already

Reply:
oh sorry I missed it